cmd/registry/cmd/upload/artifact.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/spf13/cobra"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+func artifactCommand(ctx context.Context) *cobra.Command {
+	var parent string
+	cmd := &cobra.Command{
+		Use:   "artifact FILE_PATH --parent=value",
+		Short: "Upload an artifact",
+		Args:  cobra.MinimumNArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			artifactPath := args[0]
+			if artifactPath == "" {
+				log.Fatal(ctx, "Please provide a FILE_PATH for an artifact")
+			}
+			artifact, artifactID, err := buildArtifact(ctx, artifactPath)
+			if err != nil {
+				log.FromContext(ctx).WithError(err).Fatal("Failed to read artifact")
+			}
+			client, err := connection.NewClient(ctx)
+			if err != nil {
+				log.FromContext(ctx).WithError(err).Fatal("Failed to get client")
+			}
+			artifact.Name = fmt.Sprintf("%s/artifacts/%s", parent, artifactID)
+			log.Debugf(ctx, "Uploading %s", artifact.Name)
+			err = core.SetArtifact(ctx, client, artifact)
+			if err != nil {
+				log.FromContext(ctx).WithError(err).Fatal("Failed to save artifact")
+			}
+		},
+	}
+	cmd.Flags().StringVar(&parent, "parent", "", "Parent resource for the artifact")
+	_ = cmd.MarkFlagRequired("parent")
+	return cmd
+}
+
+func buildArtifact(ctx context.Context, filename string) (*rpc.Artifact, string, error) {
+	yamlBytes, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// get the kind and id from the YAML
+	type ArtifactHeader struct {
+		Id   string `yaml:"id"`
+		Kind string `yaml:"kind"`
+	}
+	var header ArtifactHeader
+	err = yaml.Unmarshal(yamlBytes, &header)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// read the specified artifact type
+	jsonBytes, _ := yaml.YAMLToJSON(yamlBytes) // to use protojson.Unmarshal()
+	var artifact *rpc.Artifact
+	switch header.Kind {
+	case "Manifest", "google.cloud.apigeeregistry.v1.controller.Manifest":
+		artifact, err = buildManifestArtifact(ctx, jsonBytes)
+	case "TaxonomyList", "google.cloud.apigeeregistry.v1.apihub.TaxonomyList":
+		artifact, err = buildTaxonomyListArtifact(ctx, jsonBytes)
+	case "Lifecycle", "google.cloud.apigeeregistry.v1.apihub.Lifecycle":
+		artifact, err = buildLifecycleArtifact(ctx, jsonBytes)
+	default:
+		err = fmt.Errorf("unsupported artifact type %s", header.Kind)
+	}
+
+	if err != nil {
+		return nil, header.Id, err
+	}
+	return artifact, header.Id, nil
+}
+
+func buildManifestArtifact(ctx context.Context, jsonBytes []byte) (*rpc.Artifact, error) {
+	m := &rpc.Manifest{}
+	err := protojson.Unmarshal(jsonBytes, m)
+	if err != nil {
+		return nil, err
+	}
+	err = validateManifest(ctx, m)
+	if err != nil {
+		return nil, err
+	}
+	artifactBytes, err := proto.Marshal(m)
+	if err != nil {
+		return nil, err
+	}
+	return &rpc.Artifact{
+		Contents: artifactBytes,
+		MimeType: core.MimeTypeForMessageType("google.cloud.apigeeregistry.v1.controller.Manifest"),
+	}, nil
+}
+
+func validateManifest(ctx context.Context, m *rpc.Manifest) error {
```

Comment:
I think this belongs in the controller package. Does it have validation code that can be exported and reused, perhaps after some minor refactoring?

Reply:
I can look at this while adding error tests for manifest upload.

Reply:
SGTM - thanks @shrutiparabgoogle